File: RMLoaderDisplay/Components/RMDownloadIndicator.m
//
//  RMDownloadIndicator.m
//  BezierLoaders
//
//  Created by Mahesh on 1/30/14.
//  Copyright (c) 2014 Mahesh. All rights reserved.
//

#import "RMDownloadIndicator.h"
#import "RMDisplayLabel.h"

@interface RMDownloadIndicator()

// this contains list of paths to be animated through
@property(nonatomic, strong)NSMutableArray *paths;

// the shaper layers used for display
@property(nonatomic, strong)CAShapeLayer *indicateShapeLayer;
@property(nonatomic, strong)CAShapeLayer *coverLayer;

// this is the layer used for animation
@property(nonatomic, strong)CAShapeLayer *animatingLayer;

// the type of indicator
@property(nonatomic, assign)RMIndicatorType type;

// this applies to the covering stroke (default: 2)
@property(nonatomic, assign)CGFloat coverWidth;

// the last updatedPath
@property(nonatomic, strong)UIBezierPath *lastUpdatedPath;
@property(nonatomic, assign)CGFloat lastSourceAngle;

// this the animation duration (default: 0.5)
@property(nonatomic, assign)CGFloat animationDuration;


// this is display label that displays % downloaded
@property(nonatomic, strong)RMDisplayLabel *displayLabel;

@end

@implementation RMDownloadIndicator

- (id)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        // Initialization code
        _type = kRMFilledIndicator;
        [self initAttributes];
    }
    return self;
}

- (id)initWithFrame:(CGRect)frame type:(RMIndicatorType)type
{
    self = [super initWithFrame:frame];
    if (self) {
        // Initialization code
        _type = type;
        [self initAttributes];
    }
    return self;
}

/*
// Only override drawRect: if you perform custom drawing.
// An empty implementation adversely affects performance during animation.
- (void)drawRect:(CGRect)rect
{
    // Drawing code
}
*/

- (void)initAttributes
{
    // first set the radius percent attribute
    if(_type == kRMClosedIndicator)
    {
        self.radiusPercent = 0.5;
        _coverLayer = [CAShapeLayer layer];
        _animatingLayer = _coverLayer;
        
        // set the fill color
        _fillColor = [UIColor clearColor];
        _strokeColor = [UIColor whiteColor];
        _closedIndicatorBackgroundStrokeColor = [UIColor grayColor];
        _coverWidth = 2.0;
        
        //[self addDisplayLabel];
    }
    else
    {
        if(_type == kRMFilledIndicator)
        {
            // only indicateShapeLayer
            _indicateShapeLayer = [CAShapeLayer layer];
            _animatingLayer = _indicateShapeLayer;
            self.radiusPercent = 0.5;
            _coverWidth = 2.0;
        }
        else
        {
            // indicateShapeLayer and coverLayer
            _indicateShapeLayer = [CAShapeLayer layer];
            _coverLayer = [CAShapeLayer layer];
            _animatingLayer = _indicateShapeLayer;
            _coverWidth = 2.0;
            self.radiusPercent = 0.4;
        }
        
        // set the fill color
        _fillColor = [UIColor whiteColor];
        _strokeColor = [UIColor whiteColor];
        _closedIndicatorBackgroundStrokeColor = [UIColor clearColor];
    }
    
    _animatingLayer.frame = self.bounds;
    [self.layer addSublayer:_animatingLayer];
    
    // path array
    _paths = [NSMutableArray array];
    
    // animation duration
    _animationDuration = 0.5;
}

- (void)addDisplayLabel
{
    self.displayLabel = [[RMDisplayLabel alloc] initWithFrame:CGRectMake((CGRectGetWidth(self.bounds)/2 - 30/2), (CGRectGetHeight(self.bounds)/2 - 30/2), 30, 30)];
    self.displayLabel.backgroundColor = [UIColor clearColor];
    self.displayLabel.font = [UIFont fontWithName:@"HelveticaNeue-Bold" size:11.5];
    self.displayLabel.text = @"0";
    self.displayLabel.textColor = [UIColor grayColor];
    self.displayLabel.textAlignment = NSTextAlignmentCenter;
    [self addSubview:self.displayLabel];
}

- (void)loadIndicator
{
    // set the initial Path
    CGPoint center = CGPointMake(self.bounds.size.width / 2, self.bounds.size.height / 2);
    UIBezierPath *initialPath = [UIBezierPath bezierPath]; //empty path
    
    if(_type == kRMClosedIndicator)
    {
        [initialPath addArcWithCenter:center radius:(MIN(CGRectGetWidth(self.bounds), CGRectGetHeight(self.bounds))) startAngle:degreeToRadian(-90) endAngle:degreeToRadian(-90) clockwise:YES]; //add the arc
    }
    else
    {
        if(_type == kRMMixedIndictor)
        {
            [self setNeedsDisplay];
        }
        CGFloat radius = (MIN(CGRectGetWidth(self.bounds), CGRectGetHeight(self.bounds))/2) * self.radiusPercent;
        [initialPath addArcWithCenter:center radius:radius startAngle:degreeToRadian(-90) endAngle:degreeToRadian(-90) clockwise:YES]; //add the arc
    }
    
    _animatingLayer.path = initialPath.CGPath;
    _animatingLayer.strokeColor = _strokeColor.CGColor;
    _animatingLayer.fillColor = _fillColor.CGColor;
    _animatingLayer.lineWidth = _coverWidth;
    self.lastSourceAngle = degreeToRadian(-90);
}

#pragma mark -
#pragma mark Helper Methods
- (NSArray *)keyframePathsWithDuration:(CGFloat)duration lastUpdatedAngle:(CGFloat)lastUpdatedAngle newAngle:(CGFloat)newAngle radius:(CGFloat)radius type:(RMIndicatorType)type
{
    NSUInteger frameCount = ceil(duration * 60);
    NSMutableArray *array = [NSMutableArray arrayWithCapacity:frameCount + 1];
    for (int frame = 0; frame <= frameCount; frame++)
    {
        CGFloat startAngle = degreeToRadian(-90);
        CGFloat endAngle = lastUpdatedAngle + (((newAngle - lastUpdatedAngle) * frame) / frameCount);
        
        [array addObject:(id)([self pathWithStartAngle:startAngle endAngle:endAngle radius:radius type:type].CGPath)];
    }
    
    return [NSArray arrayWithArray:array];
}

- (UIBezierPath *)pathWithStartAngle:(CGFloat)startAngle endAngle:(CGFloat)endAngle radius:(CGFloat)radius type:(RMIndicatorType)type
{
    BOOL clockwise = startAngle < endAngle;
    
    UIBezierPath *path = [UIBezierPath bezierPath];
    CGPoint center = CGPointMake(self.bounds.size.width / 2, self.bounds.size.height / 2);
    
    if(type == kRMClosedIndicator)
    {
        [path addArcWithCenter:center radius:radius startAngle:startAngle endAngle:endAngle clockwise:clockwise];
    }
    else
    {
        [path moveToPoint:center];
        [path addArcWithCenter:center radius:radius startAngle:startAngle endAngle:endAngle clockwise:clockwise];
        [path closePath];
    }
    return path;
}

- (void)drawRect:(CGRect)rect
{
    if(_type == kRMMixedIndictor)
    {
        CGFloat radius = MIN(CGRectGetWidth(self.bounds), CGRectGetHeight(self.bounds))/2 - self.coverWidth;
        
        CGPoint center = CGPointMake(self.bounds.size.width / 2, self.bounds.size.height / 2);
        UIBezierPath *coverPath = [UIBezierPath bezierPath]; //empty path
        [coverPath setLineWidth:_coverWidth];
        [coverPath addArcWithCenter:center radius:radius startAngle:0 endAngle:2*M_PI clockwise:YES]; //add the arc
        
        [_strokeColor set];
        [coverPath stroke];
    }
    else if (_type == kRMClosedIndicator)
    {
        CGFloat radius = (MIN(CGRectGetWidth(self.bounds), CGRectGetHeight(self.bounds))/2) - self.coverWidth;
        
        CGPoint center = CGPointMake(self.bounds.size.width / 2, self.bounds.size.height / 2);
        UIBezierPath *coverPath = [UIBezierPath bezierPath]; //empty path
        [coverPath setLineWidth:_coverWidth];
        [coverPath addArcWithCenter:center radius:radius startAngle:0 endAngle:2*M_PI clockwise:YES]; //add the arc
        [_closedIndicatorBackgroundStrokeColor set];
        [coverPath setLineWidth:self.coverWidth];
        [coverPath stroke];
    }
}

#pragma mark - update indicator
- (void)updateWithTotalBytes:(CGFloat)bytes downloadedBytes:(CGFloat)downloadedBytes
{
    _lastUpdatedPath = [UIBezierPath bezierPathWithCGPath:_animatingLayer.path];
    
    [_paths removeAllObjects];
    
    CGFloat destinationAngle = [self destinationAngleForRatio:(downloadedBytes/bytes)];
    CGFloat radius = (MIN(CGRectGetWidth(self.bounds), CGRectGetHeight(self.bounds)) * _radiusPercent) - self.coverWidth;
    [_paths addObjectsFromArray:[self keyframePathsWithDuration:self.animationDuration lastUpdatedAngle:self.lastSourceAngle newAngle:destinationAngle  radius:radius type:_type]];
    
    _animatingLayer.path = (__bridge CGPathRef)((id)_paths[(_paths.count -1)]);
    self.lastSourceAngle = destinationAngle;
    
    CAKeyframeAnimation *pathAnimation = [CAKeyframeAnimation animationWithKeyPath:@"path"];
    [pathAnimation setValues:_paths];
    [pathAnimation setDuration:self.animationDuration];
    [pathAnimation setTimingFunction:[CAMediaTimingFunction functionWithName:kCAMediaTimingFunctionEaseOut]];
    [pathAnimation setRemovedOnCompletion:YES];
    [_animatingLayer addAnimation:pathAnimation forKey:@"path"];
    
    //[self.displayLabel updateValue:downloadedBytes/bytes];
}

- (CGFloat)destinationAngleForRatio:(CGFloat)ratio
{
    return (degreeToRadian((360*ratio) - 90));
}

float degreeToRadian(float degree)
{
    return ((degree * M_PI)/180.0f);
}

#pragma mark -
#pragma mark Setter Methods
- (void)setFillColor:(UIColor *)fillColor
{
    if(_type == kRMClosedIndicator)
        _fillColor = [UIColor clearColor];
    else
        _fillColor = fillColor;
}

- (void)setRadiusPercent:(CGFloat)radiusPercent
{
    if(_type == kRMClosedIndicator)
    {
        _radiusPercent = 0.5;
        return;
    }
    
    if(radiusPercent > 0.5 || radiusPercent < 0)
        return;
    else
        _radiusPercent = radiusPercent;
        
}

- (void)setIndicatorAnimationDuration:(CGFloat)duration
{
    self.animationDuration = duration;
}

@end

File: RMDownloadIndicator-Swift/RMDownloadIndicator-Swift/RMDownloadIndicator.swift
//
//  RMDownloadIndicator.swift
//  RMDownloadIndicator-Swift
//
//  Created by Mahesh Shanbhag on 10/08/15.
//  Copyright (c) 2015 Mahesh Shanbhag. All rights reserved.
//

import UIKit

enum RMIndicatorType: Int {
    case kRMClosedIndicator = 0
    case kRMFilledIndicator
    case kRMMixedIndictor
}

class RMDownloadIndicator: UIView {
    
    // this value should be 0 to 0.5 (default: (kRMFilledIndicator = 0.5), (kRMMixedIndictor = 0.4))
    var radiusPercent: CGFloat = 0.5 {
        didSet {
            if type == RMIndicatorType.kRMClosedIndicator {
                self.radiusPercent = 0.5
            }
            if radiusPercent > 0.5 || radiusPercent < 0 {
                radiusPercent = oldValue
            }
        }
    }
    
    // used to fill the downloaded percent slice (default: (kRMFilledIndicator = white), (kRMMixedIndictor = white))
    var fillColor: UIColor = UIColor.clearColor() {
        didSet {
            if type == RMIndicatorType.kRMClosedIndicator {
                fillColor = UIColor.clearColor()
            }
        }
    }
    
    // used to stroke the covering slice (default: (kRMClosedIndicator = white), (kRMMixedIndictor = white))
    var strokeColor: UIColor = UIColor.whiteColor()
    
    // used to stroke the background path the covering slice (default: (kRMClosedIndicator = gray))
    var closedIndicatorBackgroundStrokeColor: UIColor = UIColor.whiteColor()
    
    
    // Private properties
    private var paths: [CGPath] = []
    private var indicateShapeLayer: CAShapeLayer!
    private var coverLayer: CAShapeLayer!
    private var animatingLayer: CAShapeLayer!
    private var type: RMIndicatorType!
    private var coverWidth: CGFloat = 0.0
    private var lastUpdatedPath: UIBezierPath!
    private var lastSourceAngle: CGFloat = 0.0
    private var animationDuration: CGFloat = 0.0
    
    
    
    // init with frame and type
    // if() - (id)initWithFrame:(CGRect)frame is used the default type = kRMFilledIndicator
    override init(frame:CGRect) {
        super.init(frame:frame)
        
        self.type = .kRMFilledIndicator
        self.initAttributes()
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    init(rectframe: CGRect, type: RMIndicatorType) {
        super.init(frame: rectframe)
        
        self.type = type
        self.initAttributes()
    }
    
    func initAttributes() {
        if type == RMIndicatorType.kRMClosedIndicator {
            self.radiusPercent = 0.5
            coverLayer = CAShapeLayer()
            animatingLayer = coverLayer
            fillColor = UIColor.clearColor()
            strokeColor = UIColor.whiteColor()
            closedIndicatorBackgroundStrokeColor = UIColor.grayColor()
            coverWidth = 2.0
        }
        else {
            if type == RMIndicatorType.kRMFilledIndicator {
                indicateShapeLayer = CAShapeLayer()
                animatingLayer = indicateShapeLayer
                radiusPercent = 0.5
                coverWidth = 2.0
                closedIndicatorBackgroundStrokeColor = UIColor.clearColor()
            }
            else {
                indicateShapeLayer = CAShapeLayer()
                coverLayer = CAShapeLayer()
                animatingLayer = indicateShapeLayer
                coverWidth = 2.0
                radiusPercent = 0.4
                closedIndicatorBackgroundStrokeColor = UIColor.whiteColor()
            }
            fillColor = UIColor.whiteColor()
            strokeColor = UIColor.whiteColor()
        }
        
        animatingLayer.frame = self.bounds
        self.layer.addSublayer(animatingLayer)
        animationDuration = 0.5
    }
    
    // prepare the download indicator
    func loadIndicator() {
        let center: CGPoint = CGPointMake(self.bounds.size.width / 2, self.bounds.size.height / 2)
        let initialPath: UIBezierPath = UIBezierPath.init()
        if type == RMIndicatorType.kRMClosedIndicator {
            initialPath.addArcWithCenter(center, radius: (fmin(CGRectGetWidth(self.bounds), CGRectGetHeight(self.bounds))), startAngle: degreeToRadian(-90), endAngle: degreeToRadian(-90), clockwise: true)
        }
        else {
            if type == RMIndicatorType.kRMMixedIndictor {
                self.setNeedsDisplay()
            }
            let radius: CGFloat = (fmin(CGRectGetWidth(self.bounds), CGRectGetHeight(self.bounds)) / 2) * self.radiusPercent
            initialPath.addArcWithCenter(center, radius: radius, startAngle: degreeToRadian(-90), endAngle: degreeToRadian(-90), clockwise: true)
        }
        animatingLayer.path = initialPath.CGPath
        animatingLayer.strokeColor = strokeColor.CGColor
        animatingLayer.fillColor = fillColor.CGColor
        animatingLayer.lineWidth = coverWidth
        self.lastSourceAngle = degreeToRadian(-90)
    }
    
    func keyframePathsWithDuration(duration: CGFloat, lastUpdatedAngle: CGFloat, newAngle: CGFloat, radius: CGFloat, type: RMIndicatorType) -> [CGPath] {
        let frameCount: Int = Int(ceil(duration * 60))
        var array: [CGPath] = []
        for var frame = 0; frame <= frameCount; frame++ {
            let startAngle = degreeToRadian(-90)
            
            let angleChange = ((newAngle - lastUpdatedAngle) * CGFloat(frame))
            let endAngle = lastUpdatedAngle + (angleChange / CGFloat(frameCount))
            array.append((self.pathWithStartAngle(startAngle, endAngle: endAngle, radius: radius, type: type).CGPath))
        }
        return array
    }
    
    func pathWithStartAngle(startAngle: CGFloat, endAngle: CGFloat, radius: CGFloat, type: RMIndicatorType) -> UIBezierPath {
        let clockwise: Bool = startAngle < endAngle
        let path: UIBezierPath = UIBezierPath()
        let center: CGPoint = CGPointMake(self.bounds.size.width / 2, self.bounds.size.height / 2)
        if type == RMIndicatorType.kRMClosedIndicator {
            path.addArcWithCenter(center, radius: radius, startAngle: startAngle, endAngle: endAngle, clockwise: clockwise)
        }
        else {
            path.moveToPoint(center)
            path.addArcWithCenter(center, radius: radius, startAngle: startAngle, endAngle: endAngle, clockwise: clockwise)
            path.closePath()
        }
        return path
    }
    
    override func drawRect(rect: CGRect) {
        if type == RMIndicatorType.kRMMixedIndictor {
            let radius: CGFloat = fmin(CGRectGetWidth(self.bounds), CGRectGetHeight(self.bounds)) / 2 - self.coverWidth
            let center: CGPoint = CGPointMake(self.bounds.size.width / 2, self.bounds.size.height / 2)
            let coverPath: UIBezierPath = UIBezierPath()
            coverPath.lineWidth = coverWidth
            coverPath.addArcWithCenter(center, radius: radius, startAngle: CGFloat(0), endAngle: CGFloat(2 * M_PI), clockwise: true)
            closedIndicatorBackgroundStrokeColor.set()
            coverPath.stroke()
        }
        else {
            if type == RMIndicatorType.kRMClosedIndicator {
                let radius: CGFloat = (fmin(CGRectGetWidth(self.bounds), CGRectGetHeight(self.bounds)) / 2) - self.coverWidth
                let center: CGPoint = CGPointMake(self.bounds.size.width / 2, self.bounds.size.height / 2)
                let coverPath: UIBezierPath = UIBezierPath()
                coverPath.lineWidth = coverWidth
                coverPath.addArcWithCenter(center, radius: radius, startAngle: CGFloat(0), endAngle: CGFloat(2 * M_PI), clockwise: true)
                closedIndicatorBackgroundStrokeColor.set()
                coverPath.lineWidth = self.coverWidth
                coverPath.stroke()
            }
        }
    }
    
    // update the downloadIndicator
    func updateWithTotalBytes(bytes: CGFloat, downloadedBytes: CGFloat) {
        lastUpdatedPath = UIBezierPath.init(CGPath: animatingLayer.path!)
        paths.removeAll(keepCapacity: false)
        let destinationAngle: CGFloat = self.destinationAngleForRatio((downloadedBytes / bytes))
        let radius: CGFloat = (fmin(CGRectGetWidth(self.bounds), CGRectGetHeight(self.bounds)) * radiusPercent) - self.coverWidth
        paths = self.keyframePathsWithDuration(self.animationDuration, lastUpdatedAngle: self.lastSourceAngle, newAngle: destinationAngle, radius: radius, type: type)
        animatingLayer.path = paths[(paths.count - 1)]
        self.lastSourceAngle = destinationAngle
        let pathAnimation: CAKeyframeAnimation = CAKeyframeAnimation(keyPath: "path")
        pathAnimation.values = paths
        pathAnimation.duration = CFTimeInterval(animationDuration)
        pathAnimation.timingFunction = CAMediaTimingFunction(name: kCAMediaTimingFunctionEaseOut)
        pathAnimation.removedOnCompletion = true
        animatingLayer.addAnimation(pathAnimation, forKey: "path")
        if downloadedBytes >= bytes{
            self.removeFromSuperview()
        }
    }
    
    func destinationAngleForRatio(ratio: CGFloat) -> CGFloat {
        return (degreeToRadian((360 * ratio) - 90))
    }
    
    
    func degreeToRadian(degree: CGFloat) -> CGFloat
    {
        return (CGFloat(degree) * CGFloat(M_PI)) / CGFloat(180.0);
    }
//    
//    func setLayerFillColor(fillColor: UIColor) {
//        if type == RMIndicatorType.kRMClosedIndicator {
//            self.fillColor = UIColor.clearColor()
//        }
//        else {
//            self.fillColor = fillColor
//        }
//    }
    
//    func setLayerRadiusPercent(radiusPercent: CGFloat) {
//        if type == RMIndicatorType.kRMClosedIndicator {
//            self.radiusPercent = 0.5
//            return
//        }
//        if radiusPercent > 0.5 || radiusPercent < 0 {
//            return
//        }
//        else {
//            self.radiusPercent = radiusPercent
//        }
//    }
    
    // update the downloadIndicator
    func setIndicatorAnimationDuration(duration: CGFloat) {
        self.animationDuration = duration
    }
}
